### ISSUE TITLE
Cannot pull from Google Artifact Registry and no update on pull state

### ISSUE THREAD

**Issue by jiayihu** [deislabs/krustlet#469]
I have the hello-wasi example working if the image is pulled from GCP Registry but not from the new Artifact Registry. From logs, I can see that it panics here at line 152 with unexpected `None`:

https://github.com/deislabs/krustlet/blob/0a10d3f1e0c6ae9a2e00f61271db8c819321e465/crates/oci-distribution/src/client.rs#L143-L152

The pod YAML is just:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: hello-wasi
spec:
  containers:
    - name: hello-wasi
      image: europe-west1-docker.pkg.dev/fedra-thesis/images/hello-wasi:latest
    # (working)  image: eu.gcr.io/fedra-thesis/hello-wasi:latest
  nodeSelector:
    kubernetes.io/arch: "wasm32-wasi"
  tolerations:
    - key: "kubernetes.io/arch"
      operator: "Equal"
      value: "wasm32-wasi"
      effect: "NoExecute"
    - key: "kubernetes.io/arch"
      operator: "Equal"
      value: "wasm32-wasi"
      effect: "NoSchedule"
```

You can try it yourself to see that the pod remains in `Pending` status. I'll leave the image public, let me know if I delete it by mistake.  FYI `wasm-to-oci` is able to pull the image without issues.

Krustlet should also post updates about the image pull status, it wasn't intuitive debugging the cause of the eternal pending. The `describe` command only stated the following.

```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  19m   default-scheduler  Successfully assigned default/hello-wasi to rasp-4
```

The usual kubelet is much richer with pull events as you know:

```
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  10s               default-scheduler  Successfully assigned default/hello-world to rasp-2
  Normal   Pulled     6s                kubelet            Successfully pulled image "hello-world" in 4.094783929s
  Normal   Pulling    5s (x2 over 10s)  kubelet            Pulling image "hello-world"
  Normal   Pulled     3s                kubelet            Successfully pulled image "hello-world" in 1.714372192s
  Normal   Created    3s (x2 over 6s)   kubelet            Created container hello-world
  Normal   Started    3s (x2 over 5s)   kubelet            Started container hello-world
  Warning  BackOff    1s (x2 over 2s)   kubelet            Back-off restarting failed container
```

I haven't tried yet with the latest commit in https://github.com/deislabs/krustlet/commit/a61a59091330dc600341deb81ace2d2520973775 but I don't think it solves the issue by looking at the diff.

**Comment by jiayihu** [comment]
I may have a look at it if nobody else is working on it since I've already had a first quick read of all the source code in this project (which was very intuitive btw). But https://github.com/deislabs/krustlet/commit/a61a59091330dc600341deb81ace2d2520973775 seems to suggest that @brooksmtownsend is already working on a refactor/fix

**Comment by radu-matei** [comment]
#453 has been merged a couple of weeks ago - let us know if this is still an issue with that version (once there is a new release, or if you can build it yourself).

**Comment by jiayihu** [comment]
I already had a look at the changes in https://github.com/deislabs/krustlet/commit/a61a59091330dc600341deb81ace2d2520973775 but they are unfortunately relative to Push, not the Pull behaviour. In particular this is an issue with the authentication method, which has been changed in that PR.

Anyway, just to be sure, I recompiled from master and it results in the same error at https://github.com/deislabs/krustlet/blob/bbfd22907d609070bac337eb6913f19c0678323e/crates/oci-distribution/src/client.rs#L269

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /project/crates/oci-distribution/src/client.rs:269:50
```

Only the line number has changed, correctly pointing to the new one after the PR, meaning that I correctly updated the binary but it's not working yet. (You never know will all the things going on between Rust, cross-compilation inside Docker, Ansible etc. :D)
